src/taipy/gui/gui.py

```diff
@@ -1186,22 +1190,31 @@ def __init_route(self):
                 try:
                     self._call_function_with_state(self.on_init, [])
                 except Exception as e:
-                    warnings.warn(f"Exception on on_init execution \n{e}")
+                    warnings.warn(f"Exception on on_init execution\n{e}")
         return self._render_route()
 
     def __render_page(self, page_name: str) -> t.Any:
         self.__set_client_id_in_context()
-        page = next((page_i for page_i in self._config.pages if page_i._route == page_name), None)
+        nav_page = page_name
+        if hasattr(self, "on_navigate") and callable(self.on_navigate):
+            try:
+                nav_page = self.on_navigate(self.__get_state(), page_name)
+                if not isinstance(nav_page, str):
+                    warnings.warn(f"on_navigate returned a invalid page name '{nav_page}'")
+                    nav_page = page_name
+            except Exception as e:
+                warnings.warn(f"Exception on on_navigate execution\n{e}")
+        page = next((page_i for page_i in self._config.pages if page_i._route == nav_page), None)
```

Comment:
If nav_page != page_name, you should I think call 'on_navigate' again...

Reply:
Why would I do this ?
if new_page != page the dev changed it in on_navigate ... If I call again on_navigate, how can the dev differentiate from a "normal" call ?

Reply:
In my mind, it was to simplify the user's code, if there are several tests to perform before choosing a page name or another.
When I have a real use case, I'll let you know. Forget it for the time being.
Thx